jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/SubscriptionResource.java

```diff
@@ -216,9 +216,10 @@ public Entitlement doOperation(final CallContext ctx) throws InterruptedExceptio
                 final LocalDate resolvedEntitlementDate = requestedDate != null ? toLocalDate(requestedDate) : toLocalDate(entitlementDate);
                 final LocalDate resolvedBillingDate = requestedDate != null ? toLocalDate(requestedDate) : toLocalDate(billingDate);
                 final List<PlanPhasePriceOverride> overrides = PhasePriceOverrideJson.toPlanPhasePriceOverrides(entitlement.getPriceOverrides(), spec, account.getCurrency());
-                final Entitlement result = createAddOnEntitlement ?
-                                           entitlementApi.addEntitlement(getBundleIdForAddOnCreation(entitlement), spec, overrides, resolvedEntitlementDate, resolvedBillingDate, isMigrated, pluginProperties, callContext) :
-                                           entitlementApi.createBaseEntitlement(account.getId(), spec, entitlement.getExternalKey(), overrides, resolvedEntitlementDate, resolvedBillingDate, isMigrated, renameKeyIfExistsAndUnused, pluginProperties, callContext);
+                final Entitlement result = createAddOnEntitlement || (ProductCategory.STANDALONE.toString().equals(entitlement.getProductCategory()) && entitlement.getBundleId() != null)
```

Comment:
Nit: add parens around `(ProductCategory.STANDALONE.toString().equals(entitlement.getProductCategory()) && entitlement.getBundleId() != null)` for clarity (or maybe even extract this as `final boolean createStandaloneEntitlement`).

Reply:
I would suggest to keep the same thing with do for `ADD_ON`, where we don't mix setting boolean value `createStandaloneEntitlement ` with validation of arguments:
```

final boolean createStandaloneEntitlement = ProductCategory.STANDALONE.toString().equals(entitlement.getProductCategory());
if (createStandaloneEntitlement) {
            Preconditions.checkArgument(entitlement.getExternalKey() != null || entitlement.getBundleId() != null, "SubscriptionJson bundleId or externalKey should be specified for STANDALONE");
        }
```

In the end, it seems like `Precondition` could be shared between `STANDALONE`  and `ADD_ON` ?

Reply:
I am even wondering if one variable `isStandaloneOrAddOnEntitlement` would not be enough since they seem to share validation and end up calling same `EntitlementApi.addEntitlement` api?